### ISSUE TITLE
Arithmetic with Celsius

### ISSUE THREAD

**Issue by cousteaulecommandant** [Qalculate/libqalculate#13]
On 0.9.7, `1°C+1°C` yielded `2°C` (obviously), which converted to Kelvin gave `275.15K`.  In other words, if you have a temperature of 1 °C and raise it by 1 °C you get 2 °C.
(On the other hand, `1K+1°C` yielded `275.15K`, which doesn't fit with the description I used.)

On newer versions, however, °C is converted at parse time (which is a bit weird IMO).  This results in `1°C+1°C` being interpreted as `274.15K+274.15K` and thus yielding `548.3K` (i.e. 275.15 °C).

I'm aware of the headache that must be dealing with °C and °F programmatically (I've been in that situation before), but I'd suggest fixing this so that it makes more sense.  A fix I can think of is to only add the offset to the FIRST occurrence of a temperature, but leave the rest without the extra 273.15K.  That way only the first occurrence would be interpreted as an absolute temperature, while the rest would be treated as increments.  (This breaks the important concept of commutativity of addition though.)

Additionally, it would be nice if arithmetic done solely with °C yielded a result in °C, not K.

-----

And now if you're curious, this is how I solved this issue back in the day when I had to do a similar thing:
- Quantities can be interpreted either as absolute or differential.  `1°C` means either an absolute temperature of 274.15K or an increment (difference) of 1K.
- Quantities may have an "offset" attribute; for example, a quantity in °C has an equivalence of 1°C=1K and an offset of 273.15K.  Quantities internally keep this offset attribute during operations, along with their value.
- Adding 2 quantities preserves the padding of the first one: it assumes the first quantity is absolute and the second one represents an increment.  Thus, 1°C+2°C = 1°C+2K = 3°C = 276.15K
- Subtracting 2 quantities preserves the padding of the first quantity as well, so it's not suited for calculating a difference between absolute temperatures: it assumes the second quantity is an increment, not an absolute quantity.  (Maybe this wasn't a good idea.)
- Any other operation on quantities (multiplication, division, negation...) discards  the offset (it assumes it refers to an increment; for example °C/m means the same as K/m, and cal/°C means the same as cal/K).
I'm not sure if this "extra attribute" thing can be easily implemented with the current libqalculate; probably not.  But hopefully making the parser behave like this won't be that hard.

**Comment by hanna-kn** [comment]
Addition and subtraction (if I remember it correctly) of equal temperature units did in old versions return the result expected by most, but other calculations were problematic. My number one priority was to make the calculations mathematically correct and the handling of temperature units transparent to the user, even though the current solution has obvious flaws.

Your suggestions seem sound, but the implementation will likely require special handling and tracking of temperature units in many places in the code (and it still does not solve combined Fahrenheit and Celsius addition/subtraction). 

Because of the limited practical use (I expect most user to just enter "1+1" instead of "1°C+1°C"), I have not been motivated to spend much effort on solving the issue.

**Comment by cousteaulecommandant** [comment]
My approach would work with combined °C and °F though:
1°C+1°F = (1K+273.15K)+1°R = 274.15K+0.56K = 274.71K = 1.56°C
1°F+1°C = (1°R+459.67°R)+1K = 460.67°R+1.8°R = 462.47°R = 2.8°F
The idea is basically to only apply the "offset" to the FIRST term of the expression, and leave the rest as 1°C=1K, 1°F=1°R.

Then again, I guess I should just do 1°C+1K rather than this weird 1°C+1°C, although that feels a bit weird.  In any case, I think I should get the temperature in °C.

**Comment by hanna-kn** [comment]
Your suggestion would mean that 1°C+1°F ≠ 1°F+1°C and imply that 1°F = 0.56°C and 1°C = 1.8°F. If we, as a practical example, divide the result by two to calculate the average temperature, we then clearly get the wrong result (0.8°C instead of -8.11°C).

1°C+1°C+1K = 3°C would be similarly problematic (implies that 1°C = 1K and gives the wrong average).
By extension, even 1°C+1°C = 2°C has issues and if (for whatever reason) the result is first converted to K we once again get the wrong average.

If you must include the unit in the calculation, then I suggest that you use (1+1)°C.

I do however agree that if only one temperature is used, the result should ideally not be converted to K by default. Until that is fixed you will have to type `(1+1)°C to °C`.

**Comment by cousteaulecommandant** [comment]
Good point; I never thought of that average example.  (And yes, in my approach 1°C+1°F ≠ 1°F+1°C, but I thought that was a minor issue.)

I guess the only way around that really works would be to be able to explicitly indicate whether I'm talking about "incremental" or "offset" temperatures, so we'd need separate "offset °C" and "incremental °C" units (...also known as °C and K).

So the only way I can think of to improve the current behavior is to display a warning message if you do operations using °C or °F other than direct conversion.  (This is, assuming the average Qalculate! user may not know what absolute temperature is.)

**Comment by jjstickel** [comment]
That is exactly the problem:  oC and oF are offset units when used to describe a temperature, but the meaning is different when describing a temperature difference. A similar problem exists for pressure, which is why engineers distinguish between gauge vs. absolute pressure (e.g., psig vs psia). When working with pressure in calculations, it is best to convert to absolute pressure, do the calculations, and then convert back to gauge if desired. Following this approach with temperature, just use specifically Kelvin and Rankine for your calculations, then convert to whatever final temperature unit you want.

Perhaps a warning when using oC and oF would be helpful. Is there a warning infrastructure in qalculate?

**Comment by cousteaulecommandant** [comment]
Funny that you mention it; back in the day I also included "atm" and "ata" as 2 different things (sea level = 0 atm = 1 ata) in the implementation... but mostly so that temperatures didn't look like the exception and this looked useful; I don't actually think that's really useful.  Personally I've never used this concept of "manometric pressure" except when inflating the wheels.

**Comment by vvs-** [comment]
Units are types, not numbers. Current syntax is ambiguous, so they should be converted explicitly. A heavy syntax like `toabs(1,℃)+torel(1,℃)` should work, but a better syntax is highly desirable.

Converting types implicitly could wreak havoc. Like e.g. such innocently looking expression `a/2^b` will crash application in exact modes.

**Comment by vvs-** [comment]
This notation looks nice to me: `1℃=1K` and `Δ1℃=273.15K+1K`.

**Comment by hanna-kn** [comment]
I have now changed the handling of temperature units (and other units without proportional linear relation). Conversion has been moved from the parsing stage to when 'meeting' a related unit during calculation. I warning , that advises against calculations with multiple temperature units, has been added.

Examples of how expressions with multiple units now are calculated:
5°C+1°C=6 °C (`to °C` no longer necessary)
5°C+1K=(5-272,15)K+1K=6 °C (using `to °C`)
5°C+1°C+1K=(5+272,15)K+(1+272,15)K+1K=280.15 °C
(5°C+1°C)+1K=6°C+1K=7 °C
1°C+1°F2≈256.92778 °C
(18°C+68°F+295.15K)/3=20 °C


> Like e.g. such innocently looking expression a/2^b will crash application in exact modes.

I'm unable to reproduce this behaviour.

**Comment by vvs-** [comment]
I'm not sure. Either it was fixed since then or I meant `2/a^b` which still crashed.

**Comment by hanna-kn** [comment]
`2/a^b` has now been fixed.
number^(negative fraktion*x) caused an endless loop (which is much more serious then just crashing when raising to the power of barn).

**Comment by Rodrigodd** [comment]
Just got bitten by this issue. I was calculating the thermal energy required to heat water as:

```
> qalc '(40°C - 20°C) * (1 cal/(g*°C)) * 1kg'
((40 celsius) − (20 celsius)) × ((1 calorie) / (gram·celsius)) × (1 kilogram) ≈ 305.2343608 J
```

The actual value should be:

```
> qalc '20K * (1 cal/(g*K)) * 1kg'
(20 kelvins) × ((1 calorie) / (gram·kelvin)) × (1 kilogram) = 83.68 kJ
```

The problem is that everything is converted to Kelvin before performing the division:

```
> 40°C - 20°C
  (40 celsius) − (20 celsius) = 20 K
> 20K / °C
  (20 kelvins) / celsius ≈ 0.07295276309
> 1K/°C
  1 kelvin/celsius ≈ 0.003647638154
> 1°C to K
  1 celsius = 274.15 K
> 1K/274.15K
  (1 kelvin) / (274.15 kelvins) ≈ 0.003647638154
```

Not sure if there is any good solution for this. Maybe not converting `°C`/`°F` to absolute values when they are being multiplied by other units? Maybe that would just makes things even more complicated and confusing.

**Comment by hanna-kn** [comment]
> Just got bitten by this issue. I was calculating the thermal energy required to heat water as:
> 
> ```
> > qalc '(40°C - 20°C) * (1 cal/(g*°C)) * 1kg'
> ((40 celsius) − (20 celsius)) × ((1 calorie) / (gram·celsius)) × (1 kilogram) ≈ 305.2343608 J
> ```
> 
> The actual value should be:
> 
> ```
> > qalc '20K * (1 cal/(g*K)) * 1kg'
> (20 kelvins) × ((1 calorie) / (gram·kelvin)) × (1 kilogram) = 83.68 kJ
> ```
> (...)
>
> Not sure if there is any good solution for this. Maybe not converting `°C`/`°F` to absolute values when they are being multiplied by other units? Maybe that would just makes things even more complicated and confusing.

The result depends on the `temperature calculation` (or `temp`) setting. The expression returns 83.68 kJ with the default `hybrid` (or `0`) option.

**Comment by Rodrigodd** [comment]
> The result depends on the temperature calculation (or temp) setting. The expression returns 83.68 kJ with the default hybrid (or 0) option.

Thanks, changing it to 0 fixed the issue. I may have selected the wrong option when it prompted me the first time.
